File: src/test/scala/com/exasol/parquetio/reader/RowParquetReaderPrimitiveTypesTest.scala
package com.exasol.parquetio.reader

import java.math.BigDecimal
import java.math.BigInteger
import java.math.MathContext
import java.nio.charset.StandardCharsets.UTF_8
import java.sql.Timestamp

import com.exasol.parquetio.data.GenericRow
import com.exasol.parquetio.BaseParquetReaderTest

import org.apache.parquet.example.data.simple.SimpleGroup
import org.apache.parquet.io.api.Binary
import org.apache.parquet.schema.MessageTypeParser

// [utest->dsn~converting-primitive-column-types~1]
// [utest->dsn~converting-logical-column-types~1]
class RowParquetReaderPrimitiveTypesTest extends BaseParquetReaderTest {

  test("reads INT64 (TIMESTAMP_MILLIS) as timestamp value") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required int64 col_long;
         |  required int64 col_timestamp (TIMESTAMP_MILLIS);
         |}
         |""".stripMargin
    )
    val timestamp = new Timestamp(System.currentTimeMillis())
    withResource(getParquetWriter(schema, false)) { writer =>
      val record = new SimpleGroup(schema)
      record.append("col_long", 153L)
      record.append("col_timestamp", timestamp.getTime())
      writer.write(record)
    }
    assert(getRecords() === Seq(GenericRow.of(java.lang.Long.valueOf(153), timestamp)))
  }

  test("reads INT64 (TIMESTAMP_MICROS) as timestamp value") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required int64 col_timestamp (TIMESTAMP_MICROS);
         |}
         |""".stripMargin
    )
    val timestamp = Timestamp.valueOf("2022-01-12 11:52:53.123456")
    val micros = timestamp.getTime() * 1000L + (timestamp.getNanos().toLong / 1000) % 1000L
    withResource(getParquetWriter(schema, false)) { writer =>
      val record = new SimpleGroup(schema)
      record.append("col_timestamp", micros)
      writer.write(record)
    }
    assert(getRecords() === Seq(GenericRow.of(timestamp)))
  }

  test("reads FIXED_LEN_BYTE_ARRAY as string value") {
    val size = 5
    val schema = MessageTypeParser.parseMessageType(
      s"""|message test {
          |  required fixed_len_byte_array($size) col_byte_array;
          |}
          |""".stripMargin
    )
    withResource(getParquetWriter(schema, false)) { writer =>
      val record = new SimpleGroup(schema)
      record.append("col_byte_array", "hello")
      writer.write(record)
    }
    assert(getRecords() === Seq(GenericRow.of("hello")))
  }

  test("reads BINARY as string value") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required binary col_binary;
         |}
         |""".stripMargin
    )
    withResource(getParquetWriter(schema, false)) { writer =>
      val record = new SimpleGroup(schema)
      record.append("col_binary", "test")
      writer.write(record)
    }
    assert(getRecords() === Seq(GenericRow.of("test")))
  }

  test("reads BINARY (UTF8) as string value using dictionary encoding") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required binary col_binary (UTF8);
         |}
         |""".stripMargin
    )
    withResource(getParquetWriter(schema, true)) { writer =>
      Seq("test1", "test2", "test1", "test2").foreach { value =>
        val record = new SimpleGroup(schema)
        record.append("col_binary", value)
        writer.write(record)
      }
    }
    val records = getRecords()
    assert(records.size === 4)
    assert(records(0) === GenericRow.of("test1"))
    assert(records(1) === GenericRow.of("test2"))
  }

  test("reads INT32 (decimal) as big decimal value using dictionary encoding") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required int32 col_int (DECIMAL(9,2));
         |}
         |""".stripMargin
    )
    withResource(getParquetWriter(schema, true)) { writer =>
      Seq(123456789, 0, 0, 123456789).foreach { value =>
        val record = new SimpleGroup(schema)
        record.append("col_int", value)
        writer.write(record)
      }
    }
    val records = getRecords()
    assert(records.size === 4)
    assert(records(0) === GenericRow.of(BigDecimal.valueOf(123456789, 2)))
    assert(records(1) === GenericRow.of(BigDecimal.valueOf(0, 2)))
  }

  test("reads INT64 (decimal) as big decimal value using dictionary encoding") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required int64 column (DECIMAL(18,2));
         |}
         |""".stripMargin
    )
    withResource(getParquetWriter(schema, true)) { writer =>
      Seq(1234567890123456L, 1L, 1L, 1234567890123456L).foreach { value =>
        val record = new SimpleGroup(schema)
        record.append("column", value)
        writer.write(record)
      }
    }
    val records = getRecords()
    assert(records.size === 4)
    assert(records(0) === GenericRow.of(BigDecimal.valueOf(1234567890123456L, 2)))
    assert(records(1) === GenericRow.of(BigDecimal.valueOf(1L, 2)))
  }

  test("reads BINARY (decimal) as big decimal value using dictionary encoding") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required binary column (DECIMAL(30,2));
         |}
         |""".stripMargin
    )
    val decimalValue = "123456789012345678901234567890"
    withResource(getParquetWriter(schema, true)) { writer =>
      Seq(decimalValue, decimalValue).foreach { value =>
        val record = new SimpleGroup(schema)
        record.append("column", value)
        writer.write(record)
      }
    }
    val expected = new BigDecimal(new BigInteger(decimalValue.getBytes(UTF_8)), 2, new MathContext(30))
    val records = getRecords()
    assert(records.size === 2)
    assert(records(0) === GenericRow.of(expected))
    assert(records(1) === GenericRow.of(expected))
  }

  test("reads FIXED_LEN_BYTE_ARRAY (decimal) as big decimal value using dictionary encoding") {
    val schema = MessageTypeParser.parseMessageType(
      """|message test {
         |  required fixed_len_byte_array(9) column (DECIMAL(20,2));
         |}
         |""".stripMargin
    )
    val decimalValue = "12345678901234567890"
    val decimalBinary = Binary.fromConstantByteArray(new BigDecimal(decimalValue).unscaledValue().toByteArray())
    val zeros = Binary.fromConstantByteArray(Array.fill[Byte](9)(0x0), 0, 9)
    withResource(getParquetWriter(schema, true)) { writer =>
      Seq(decimalBinary, zeros).foreach { value =>
        val record = new SimpleGroup(schema)
        record.append("column", value)
        writer.write(record)
      }
    }
    val expected = new BigDecimal(new BigInteger(decimalValue), 2, new MathContext(20))
    val records = getRecords()
    assert(records.size === 2)
    assert(records(0) === GenericRow.of(expected))
    assert(records(1) === GenericRow.of(BigDecimal.valueOf(0, 2)))
  }

}

File: src/test/java/com/exasol/parquetio/reader/RowParquetReaderPrimitiveTypesTest.java
package com.exasol.parquetio.reader;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import static org.junit.jupiter.api.Assertions.assertAll;

import java.io.IOException;
import java.math.BigDecimal;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.sql.Date;
import java.sql.Timestamp;
import java.time.Instant;
import java.util.List;
import java.util.UUID;

import org.apache.parquet.example.data.Group;
import org.apache.parquet.example.data.simple.SimpleGroup;
import org.apache.parquet.hadoop.ParquetWriter;
import org.apache.parquet.io.api.Binary;
import org.apache.parquet.schema.MessageType;
import org.junit.jupiter.api.Test;

import com.exasol.parquetio.BaseParquetReaderTest;
import com.exasol.parquetio.data.GenericRow;
import com.exasol.parquetio.data.Row;

// [utest->dsn~converting-primitive-column-types~1]
// [utest->dsn~converting-logical-column-types~1]
class RowParquetReaderPrimitiveTypesTest extends BaseParquetReaderTest {
    private static final int JULIAN_DAY_OF_EPOCH = 2_440_588;

    @Test
    void testReadsInt64TimestampMillisAsTimestampValue() throws IOException {
        final MessageType schema = parseSchema("message test {", "  required int64 col_long;",
                "  required int64 col_timestamp (TIMESTAMP_MILLIS);", "}");
        final long millisSinceEpoch = 1_712_297_228_009L;
        final Timestamp timestamp = Timestamp.from(Instant.parse("2024-04-05T06:07:08.009Z"));
        try (ParquetWriter<Group> writer = getParquetWriter(schema, false)) {
            final SimpleGroup parquetRow = new SimpleGroup(schema);
            parquetRow.append("col_long", 153L);
            parquetRow.append("col_timestamp", millisSinceEpoch);
            writer.write(parquetRow);
        }
        assertThat(getRecords(), contains(GenericRow.of(153L, timestamp)));
    }

    @Test
    void testReadsInt64TimestampMicrosAsTimestampValue() throws IOException {
        final MessageType schema = parseSchema("message test {", "  required int64 col_timestamp (TIMESTAMP_MICROS);",
                "}");
        final long microsSinceEpoch = 1_641_988_373_123_456L;
        final Timestamp timestamp = Timestamp.from(Instant.parse("2022-01-12T11:52:53.123456Z"));
        try (ParquetWriter<Group> writer = getParquetWriter(schema, false)) {
            final SimpleGroup parquetRow = new SimpleGroup(schema);
            parquetRow.append("col_timestamp", microsSinceEpoch);
            writer.write(parquetRow);
        }
        assertThat(getRecords(), contains(GenericRow.of(timestamp)));
    }

    @Test
    void testReadsInt96TimestampAsTimestampValue() throws IOException {
        final MessageType schema = parseSchema("message test {", "  required int96 col_timestamp;", "}");
        final long nanosSinceStartOfDay = 45_296_123_456_789L;
        try (ParquetWriter<Group> writer = getParquetWriter(schema, false)) {
            final SimpleGroup parquetRow = new SimpleGroup(schema);
            parquetRow.append("col_timestamp", getInt96Timestamp(JULIAN_DAY_OF_EPOCH + 1, nanosSinceStartOfDay));
            writer.write(parquetRow);
        }
        final Timestamp expected = Timestamp.from(Instant.parse("1970-01-02T12:34:56.123456Z"));
        assertThat(getRecords(), contains(GenericRow.of(expected)));
    }

    @Test
    void testReadsInt32DateAsDateValue() throws IOException {
        final MessageType schema = parseSchema("message test {", "  required int32 col_date (DATE);", "}");
        try (ParquetWriter<Group> writer = getParquetWriter(schema, false)) {
            final SimpleGroup parquetRow = new SimpleGroup(schema);
            parquetRow.append("col_date", 19_723);
            writer.write(parquetRow);
        }
        assertThat(getRecords(), contains(GenericRow.of(Date.valueOf("2024-01-01"))));
    }

    private Binary getInt96Timestamp(final int julianDay, final long nanosSinceStartOfDay) {
        final byte[] bytes = ByteBuffer.allocate(12).order(ByteOrder.LITTLE_ENDIAN).putLong(nanosSinceStartOfDay)
                .putInt(julianDay).array();
        return Binary.fromConstantByteArray(bytes);
    }

    @Test
    void testReadsFixedLenByteArrayAsStringValue() throws IOException {
        final int size = 5;
        final MessageType schema = parseSchema("message test {", "  required fixed_len_byte_array(" + size
                + ") col_byte_array;", "}");
        try (ParquetWriter<Group> writer = getParquetWriter(schema, false)) {
            final SimpleGroup parquetRow = new SimpleGroup(schema);
            parquetRow.append("col_byte_array", "hello");
            writer.write(parquetRow);
        }
        assertThat(getRecords(), contains(GenericRow.of("hello")));
    }

    @Test
    void testReadsFixedLenByteArrayUuidAsUuidValue() throws IOException {
        final MessageType schema = parseSchema("message test {", "  required fixed_len_byte_array(16) col_uuid (UUID);",
                "}");
        final UUID uuid = UUID.fromString("123e4567-e89b-12d3-a456-426614174000");
        try (ParquetWriter<Group> writer = getParquetWriter(schema, false)) {
            final SimpleGroup parquetRow = new SimpleGroup(schema);
            parquetRow.append("col_uuid", getByteBufferFromUuid(uuid));
            writer.write(parquetRow);
        }
        assertThat(getRecords(), contains(GenericRow.of(uuid)));
    }

    private Binary getByteBufferFromUuid(final UUID uuid) {
        final byte[] bytes = ByteBuffer.allocate(16).order(ByteOrder.BIG_ENDIAN)
                .putLong(uuid.getMostSignificantBits()).putLong(uuid.getLeastSignificantBits()).array();
        return Binary.fromConstantByteArray(bytes);
    }

    @Test
    void testReadsBinaryAsStringValue() throws IOException {
        final MessageType schema = parseSchema("message test {", "  required binary col_binary;", "}");
        try (ParquetWriter<Group> writer = getParquetWriter(schema, false)) {
            final SimpleGroup parquetRow = new SimpleGroup(schema);
            parquetRow.append("col_binary", "test");
            writer.write(parquetRow);
        }
        assertThat(getRecords(), contains(GenericRow.of("test")));
    }

    @Test
    void testReadsBinaryUtf8AsStringValueUsingDictionaryEncoding() throws IOException {
        final MessageType schema = parseSchema("message test {", "  required binary col_binary (UTF8);", "}");
        try (ParquetWriter<Group> writer = getParquetWriter(schema, true)) {
            for (final String value : List.of("north", "south", "north", "south")) {
                final SimpleGroup parquetRow = new SimpleGroup(schema);
                parquetRow.append("col_binary", value);
                writer.write(parquetRow);
            }
        }
        final List<Row> records = getRecords();
        assertAll(() -> assertThat(records, hasSize(4)),
                () -> assertThat(records.get(0), equalTo(GenericRow.of("north"))),
                () -> assertThat(records.get(1), equalTo(GenericRow.of("south"))));
    }

    @Test
    void testReadsInt32DecimalAsBigDecimalValueUsingDictionaryEncoding() throws IOException {
        final MessageType schema = parseSchema("message test {", "  required int32 col_int (DECIMAL(9,2));", "}");
        try (ParquetWriter<Group> writer = getParquetWriter(schema, true)) {
            for (final int value : List.of(123456789, 0, 0, 123456789)) {
                final SimpleGroup parquetRow = new SimpleGroup(schema);
                parquetRow.append("col_int", value);
                writer.write(parquetRow);
            }
        }
        final List<Row> records = getRecords();
        assertAll(() -> assertThat(records, hasSize(4)),
                () -> assertThat(records.get(0), equalTo(GenericRow.of(BigDecimal.valueOf(123456789, 2)))),
                () -> assertThat(records.get(1), equalTo(GenericRow.of(BigDecimal.valueOf(0, 2)))));
    }

    @Test
    void testReadsInt64DecimalAsBigDecimalValueUsingDictionaryEncoding() throws IOException {
        final MessageType schema = parseSchema("message test {", "  required int64 column (DECIMAL(18,2));", "}");
        try (ParquetWriter<Group> writer = getParquetWriter(schema, true)) {
            for (final long value : List.of(1234567890123456L, 1L, 1L, 1234567890123456L)) {
                final SimpleGroup parquetRow = new SimpleGroup(schema);
                parquetRow.append("column", value);
                writer.write(parquetRow);
            }
        }
        final List<Row> records = getRecords();
        assertAll(() -> assertThat(records, hasSize(4)),
                () -> assertThat(records.get(0), equalTo(GenericRow.of(BigDecimal.valueOf(1234567890123456L, 2)))),
                () -> assertThat(records.get(1), equalTo(GenericRow.of(BigDecimal.valueOf(1L, 2)))));
    }

    @Test
    void testReadsBinaryDecimalAsBigDecimalValueUsingDictionaryEncoding() throws IOException {
        final MessageType schema = parseSchema("message test {", "  required binary column (DECIMAL(30,2));", "}");
        final String decimalValue = "123456789012345678901234567890";
        try (ParquetWriter<Group> writer = getParquetWriter(schema, true)) {
            for (final String value : List.of(decimalValue, decimalValue)) {
                final SimpleGroup parquetRow = new SimpleGroup(schema);
                parquetRow.append("column", value);
                writer.write(parquetRow);
            }
        }
        final BigDecimal expected = new BigDecimal("3.39538960730037532986880915048E+69");
        final List<Row> records = getRecords();
        assertAll(() -> assertThat(records, hasSize(2)), () -> assertThat(records.get(0), equalTo(GenericRow.of(expected))),
                () -> assertThat(records.get(1), equalTo(GenericRow.of(expected))));
    }

    @Test
    void testReadsFixedLenByteArrayDecimalAsBigDecimalValueUsingDictionaryEncoding() throws IOException {
        final MessageType schema = parseSchema("message test {",
                "  required fixed_len_byte_array(9) column (DECIMAL(20,2));", "}");
        final String decimalValue = "12345678901234567890";
        final Binary decimalBinary = Binary
                .fromConstantByteArray(new BigDecimal(decimalValue).unscaledValue().toByteArray());
        final Binary zeros = Binary.fromConstantByteArray(new byte[9], 0, 9);
        try (ParquetWriter<Group> writer = getParquetWriter(schema, true)) {
            for (final Binary value : List.of(decimalBinary, zeros)) {
                final SimpleGroup parquetRow = new SimpleGroup(schema);
                parquetRow.append("column", value);
                writer.write(parquetRow);
            }
        }
        final BigDecimal expected = new BigDecimal("123456789012345678.90");
        final List<Row> records = getRecords();
        assertAll(() -> assertThat(records, hasSize(2)), () -> assertThat(records.get(0), equalTo(GenericRow.of(expected))),
                () -> assertThat(records.get(1), equalTo(GenericRow.of(BigDecimal.valueOf(0, 2)))));
    }
}
